### ISSUE TITLE
SpinLockDelay in spinlock_win32-inl.h could be rewritten with WaitOnAddress for better performance

### ISSUE THREAD

**Issue by lennoxho** [gperftools/gperftools#1333]
This snippet in spinlock_win32-inl.h
https://github.com/gperftools/gperftools/blob/fe85bbdf4cb891a67a8e2109c1c22a33aa958c7e/src/base/spinlock_win32-inl.h#L41-L51
could be rewritten with [WaitOnAddress](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress)
WaitOnAddress is equvalent to futex, which is the solution used on Linux.

```C++
void SpinLockDelay(volatile Atomic32 *w, int32 value, int loop) {
    if (loop != 0) {
        auto wait_ns = static_cast<uint64_t>(base::internal::SuggestedDelayNS(loop)) * 16ull;
        auto wait_ms = wait_ns / 1000000ull;

        WaitOnAddress(w, &value, 4u, static_cast<DWORD>(wait_ms));
    }
}

void SpinLockWake(volatile Atomic32 *w, bool all) {
    if (all) WakeByAddressAll((void*)w);
    else WakeByAddressSingle((void*)w);
}
```
In our project, we have 10 threads performing small allocations in parallel.
For this workload, we are seeing massive (10%-ish) improvement in runtime.

This VTune chart perfectly demonstrates what the change does to CPU utilisation
![image](https://user-images.githubusercontent.com/9140100/164534102-384ea964-cf14-431c-964c-3b6a97f95b19.png)
(top: before; bottom: after)

**Comment by alk** [comment]
Yes. Patches welcome. Also consider "stealing" whatever is being done in abseil. They may have modernized this too.

**Comment by shiyi23** [comment]
@alk could you asign me？
/asign

**Comment by alk** [comment]
Oh, wasn't even aware there is such feature. Sure, if you ask.

**Comment by lennoxho** [comment]
New PR attempt which addresses the comment to support Windows 7 or older.
https://github.com/gperftools/gperftools/pull/1438

**Comment by lennoxho** [comment]
New PR https://github.com/gperftools/gperftools/pull/1440
